cmd/erasure-readfile.go

```diff
@@ -23,214 +23,166 @@ import (
-	// chunk size)
-	isLastBlock := (totalLength-1)/blocksize == (offset+length-1)/blocksize
-	var partDataEndIndex int64
-	if isLastBlock {
-		lastBlockChunkSize := chunksize
-		if totalLength%blocksize != 0 {
-			lastBlockChunkSize = ceilFrac(totalLength%blocksize, int64(s.dataBlocks))
-		}
-		partDataEndIndex = blockStartIndex + lastBlockChunkSize - 1
-	} else {
-		partDataEndIndex = blockStartIndex + chunksize - 1
+	if p.offset+p.shardSize > p.shardFileSize {
+		p.shardSize = p.shardFileSize - p.offset
 	}
 
-	// Thus, the length of data to be read from the part file(s) is:
-	partDataLength := partDataEndIndex - partDataStartIndex + 1
-	// The calculation above does not apply when length == 0:
-	if length == 0 {
-		partDataLength = 0
+	read := func(currReaderIndex int) {
+		b, err := p.readers[currReaderIndex].Read(p.offset, p.shardSize)
+		errCh <- errIdx{currReaderIndex, b, err}
 	}
 
-	var buffers [][]byte
-	var needsReconstruction bool
-	buffers, needsReconstruction, err = s.readConcurrent(ctx, volume, path,
-		partDataStartIndex, partDataLength, verifiers)
-	if err != nil {
-		// Could not read enough disks.
-		return
+	readerCount := 0
+	for _, r := range p.readers {
+		if r != nil {
+			readerCount++
+		}
+	}
+	if readerCount < p.dataBlocks {
+		return nil, errXLReadQuorum
 	}
 
-	numChunks := ceilFrac(partDataLength, chunksize)
-	blocks := make([][]byte, len(s.disks))
-
-	if needsReconstruction && numChunks > 1 {
-		// Allocate once for all the equal length blocks. The
-		// last block may have a different length - allocation
-		// for this happens inside the for loop below.
-		for i := range blocks {
-			if len(buffers[i]) == 0 {
-				blocks[i] = make([]byte, chunksize)
-			}
+	readerCount = 0
+	for i, r := range p.readers {
+		if r == nil {
+			continue
+		}
+		go read(i)
+		readerCount++
+		if readerCount == p.dataBlocks {
+			currReaderIndex = i + 1
+			break
 		}
 	}
 
-	var buffOffset int64
-	for chunkNumber := int64(0); chunkNumber < numChunks; chunkNumber++ {
-		if chunkNumber == numChunks-1 && partDataLength%chunksize != 0 {
-			chunksize = partDataLength % chunksize
-			// We allocate again as the last chunk has a
-			// different size.
-			for i := range blocks {
-				if len(buffers[i]) == 0 {
-					blocks[i] = make([]byte, chunksize)
-				}
+	for errVal := range errCh {
+		if errVal.err == nil {
+			newBuf[errVal.idx] = errVal.buf
+			if p.canDecode(newBuf) {
+				p.offset += int64(p.shardSize)
+				return newBuf, nil
 			}
+			continue
 		}
-
-		for i := range blocks {
-			if len(buffers[i]) == 0 {
-				blocks[i] = blocks[i][0:0]
+		p.readers[errVal.idx] = nil
+		for currReaderIndex < len(p.readers) {
+			if p.readers[currReaderIndex] != nil {
+				break
 			}
+			currReaderIndex++
 		}
 
-		for i := range blocks {
-			if len(buffers[i]) != 0 {
-				blocks[i] = buffers[i][buffOffset : buffOffset+chunksize]
-			}
+		if currReaderIndex == len(p.readers) {
+			break
 		}
-		buffOffset += chunksize
+		go read(currReaderIndex)
```

Comment:
You span go-routines you may never wait for. These go routines try write to channel - this may causes a memory leak. Go concurrency cannot be done like that...

Reply:
No, there is no go routine leak because we would not even spawn a go routine in the following case:
```
80         readerCount := 0                                                                                                                                                                                        
81         for _, r := range p.readers {                                                                                                                                                                           
82                 if r != nil {                                                                                                                                                                                   
83                         readerCount++                                                                                                                                                                           
84                 }                                                                                                                                                                                               
85         }                                                                                                                                                                                                       
86         if readerCount < p.dataBlocks {                                                                                                                                                                         
87                 return nil, errXLReadQuorum                                                                                                                                                                     
88         }                                                                                                                                                                                                       
```

can you check it again for the situation where you think there is a go routine leak?

Reply:
I think you're actually right, since you implicitly disable the reader here `p.readers[errVal.idx] = nil`